### ISSUE TITLE
SQLite error on upsert and PIL exception

### ISSUE THREAD

**Issue by LoopControl** [neuml/txtai#386]
After indexing/upserting around 200k images, I'm now seeing this exception about an SQLite constraint error. Any suggestions to recover?

```
    embeddings.upsert(images(x, y))
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/base.py", line 158, in upsert
    ids, _, embeddings = transform(documents, buffer)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/transform.py", line 58, in __call__
    ids, dimensions, batches, stream = self.model.index(self.stream(documents), self.batch)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/vectors/base.py", line 79, in index
    for document in documents:
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/transform.py", line 112, in stream
    self.load(batch, offset)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/transform.py", line 142, in load
    self.database.insert(batch, self.offset)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/database/sqlite.py", line 151, in insert
    self.insertsection(index, uid, document, tags, entry)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/database/sqlite.py", line 433, in insertsection
    self.cursor.execute(SQLite.INSERT_SECTION, [index, uid, text, tags, entry])
sqlite3.IntegrityError: UNIQUE constraint failed: sections.indexid

Exception ignored in: <generator object images at 0x7fe384949770>
RuntimeError: generator ignored GeneratorExit
```

Current config (`embeddings.info()` output):
```
{
  "backend": "faiss",
  "batch": 2,
  "build": {
    "create": "2022-11-24T23:38:28Z",
    "python": "3.10.6",
    "settings": {
      "components": "IDMap,Flat"
    },
    "system": "Linux (x86_64)",
    "txtai": "5.1.0"
  },
  "content": true,
  "dimensions": 512,
  "encodebatch": 2,
  "method": "sentence-transformers",
  "offset": 201487,
  "path": "clip-ViT-B-32",
  "update": "2022-11-25T03:38:44Z"
}
```

**Comment by LoopControl** [comment]
Did a hacky solution to get past the issue (did a `DELETE FROM sections where indexid > 201487` which was the indexid it kept stopping on) which worked for a few hundred iterations until it failed with this error:

```
Traceback (most recent call last):
  File "/home/loop/image-search/index.py", line 68, in <module>
    embeddings.upsert(images(x, y))
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/base.py", line 158, in upsert
    ids, _, embeddings = transform(documents, buffer)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/embeddings/transform.py", line 58, in __call__
    ids, dimensions, batches, stream = self.model.index(self.stream(documents), self.batch)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/vectors/base.py", line 84, in index
    uids, dimensions = self.batch(batch, output)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/vectors/base.py", line 130, in batch
    embeddings = self.encode(documents)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/txtai/vectors/transformers.py", line 48, in encode
    return self.model.encode(data, self.encodebatch)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/sentence_transformers/SentenceTransformer.py", line 161, in encode
    features = self.tokenize(sentences_batch)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/sentence_transformers/SentenceTransformer.py", line 319, in tokenize
    return self._first_module().tokenize(texts)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/sentence_transformers/models/CLIPModel.py", line 71, in tokenize
    inputs = self.processor(text=texts_values, images=images, return_tensors="pt", padding=True)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/transformers/models/clip/processing_clip.py", line 85, in __call__
    image_features = self.feature_extractor(images, return_tensors=return_tensors, **kwargs)
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/transformers/models/clip/feature_extraction_clip.py", line 153, in __call__
    images = [self.convert_rgb(image) for image in images]
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/transformers/models/clip/feature_extraction_clip.py", line 153, in <listcomp>
    images = [self.convert_rgb(image) for image in images]
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/transformers/image_utils.py", line 246, in convert_rgb
    return image.convert("RGB")
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/PIL/Image.py", line 921, in convert
    self.load()
  File "/home/loop/apps/anaconda3/envs/pytorch/lib/python3.10/site-packages/PIL/ImageFile.py", line 254, in load
    raise OSError(
OSError: image file is truncated (19 bytes not processed)
! images() got an exception /media/66/101/BeeoVZW4.jpg
Exception ignored in: <generator object images at 0x7f61644517e0>
RuntimeError: generator ignored GeneratorExit
```

**Comment by davidmezzetti** [comment]
Regarding the PIL exception, I've never seen the error before but from what I've read, it seems to either be related to having corrupted images (but recoverable) or PIL is choking on large images. Are you working with the latest release of Pillow?

There was another response in that SO thread that said they used opencv to read images - https://stackoverflow.com/questions/12984426/pil-ioerror-image-file-truncated-with-big-images/68789661#68789661

You could consider filing an issue over with https://github.com/python-pillow/Pillow if it seems like the images should work without issue. Otherwise, sounds like one of the workarounds, will keep you moving forward.

**Comment by davidmezzetti** [comment]
The unique constraint issue is odd, also another one I haven't seen before.

Are you indexing with multiple processes/threads?

**Comment by LoopControl** [comment]
The PIL issue was resolved with the `LOAD_TRUNCATED_IMAGES`. It happened on a 200KB image that had small dimensions and seems pretty rare (likely partially corrupt JPG even though it displays fine in browser) though.

I let the script run overnight with this fix and it's happily chugging along -- a little over a million images indexed in the last 8 hours, 2.1GB embeddings file and 210MB documents file.

The constraint issue happened when I had the `upsert()` wrapped in a try/except. After PIL throws the above exception, it seems the sqlite error is thrown when you resume. Now that the truncated image in PIL is resolved, I removed the try/catch on the upsert and haven't seen the constraint error since.
